src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -188,46 +186,106 @@ private void EnqueueProcessSnapshot(Document newDocument)
                         _workQueue.AssertIsForeground();
-                        ReportChangedSpan(snapshot.GetFullSpan());
+                        ReportChangedSpan(changedSpan);
                     },
                     ReportChangeDelayInMilliseconds,
                     _listener.BeginAsyncOperation("ReportEntireFileChanged"),
                     _reportChangeCancellationSource.Token);
             }
 
+            private async Task<SnapshotSpan> GetChangedSpanAsync(
+                Document previousDocument, Document currentDocument,
+                ITextSnapshot currentSnapshot, CancellationToken cancellationToken)
+            {
+                // Try to defer to the classification service to see if it can find a narrower range to recompute.
+                if (previousDocument != null)
+                {
+                    var classificationService = TryGetClassificationService(currentSnapshot);
+                    if (classificationService != null)
+                    {
+                        var changeRange = await ComputeChangeRangeAsync(
+                            previousDocument, currentDocument, classificationService, cancellationToken).ConfigureAwait(false);
+                        if (changeRange != null)
+                            return currentSnapshot.GetSpan(changeRange.Value.Span.Start, changeRange.Value.NewLength);
+                    }
+                }
+
+                // Couldn't compute a narrower range.  Just the mark the entire file as changed.
+                return currentSnapshot.GetFullSpan();
+            }
+
+            private static async Task<TextChangeRange?> ComputeChangeRangeAsync(
+                Document previousDocument, Document currentDocument,
+                IClassificationService classificationService, CancellationToken cancellationToken)
+            {
+                // We want to compute a minimal change, but we don't want this to run for too long.  So do the
+                // computation work in the threadpool, but also gate how much time we can spend here so that we can let
+                // the editor know about the size of the change asap.
+                using var linkedToken = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+                var computeTask = Task.Run(() =>
+                    classificationService.ComputeSyntacticChangeRangeAsync(previousDocument, currentDocument, linkedToken.Token), linkedToken.Token);
+
+                var delayTask = Task.Delay(TaggerConstants.NearImmediateDelay, linkedToken.Token);
+
+                var completedTask = await Task.WhenAny(computeTask, delayTask).ConfigureAwait(false);
+
+                // Ensure that we cancel any outstanding work on the other task, once one completes.
+                linkedToken.Cancel();
+
+                // ensure that if we completed because of cancellation, we throw that up.
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
not sure about this canclelation handshake.  i always find linked tokens a bit clunky.